libs/vkd3d/memory.c

```diff
@@ -380,6 +394,197 @@ static HRESULT vkd3d_memory_allocation_init(struct vkd3d_memory_allocation *allo
+}
+
+static HRESULT vkd3d_memory_chunk_allocate_range(struct vkd3d_memory_chunk *chunk, const VkMemoryRequirements *memory_requirements,
+        struct vkd3d_memory_allocation *allocation)
+{
+    struct vkd3d_memory_free_range *pick_range;
+    VkDeviceSize l_length, r_length;
+    size_t i, pick_index;
+
+    if (!chunk->free_ranges_count)
+        return E_OUTOFMEMORY;
+
+    pick_index = chunk->free_ranges_count;
+    pick_range = NULL;
+
+    for (i = 0; i < chunk->free_ranges_count; i++)
+    {
+        struct vkd3d_memory_free_range *range = &chunk->free_ranges[i];
+
+        if (range->offset + range->length - align(range->offset, memory_requirements->alignment) < memory_requirements->size)
+            continue;
+
+        /* Exact fit leaving no gaps */
+        if (range->length == memory_requirements->size)
+        {
+            pick_index = i;
+            pick_range = range;
+            break;
+        }
+
+        /* Alignment is almost always going to be 64 KiB, so
+         * don't worry too much about misalignment gaps here */
+        if (!pick_range || range->length > pick_range->length)
+        {
+            pick_index = i;
+            pick_range = range;
+        }
+    }
+
+    if (!pick_range)
+        return E_OUTOFMEMORY;
+
+    /* Adjust offsets and addresses of the base allocation */
+    vkd3d_memory_allocation_slice(allocation, &chunk->allocation,
+            align(pick_range->offset, memory_requirements->alignment),
+            memory_requirements->size);
+
+    /* Remove allocated range from the free list */
+    l_length = allocation->offset - pick_range->offset;
+    r_length = pick_range->offset + pick_range->length
+            - allocation->offset - allocation->resource.size;
+
+    if (l_length)
+    {
+        pick_range->length = l_length;
+
+        if (r_length)
+        {
+            vkd3d_memory_chunk_insert_range(chunk, pick_index + 1,
+                allocation->offset + allocation->resource.size, r_length);
+        }
+    }
+    else if (r_length)
+    {
+        pick_range->offset = allocation->offset + allocation->resource.size;
+        pick_range->length = r_length;
+    }
+    else
+    {
+        vkd3d_memory_chunk_remove_range(chunk, pick_index);
+    }
+
+    return S_OK;
+}
+
+static size_t vkd3d_memory_chunk_find_range(struct vkd3d_memory_chunk *chunk, VkDeviceSize offset)
+{
+    struct vkd3d_memory_free_range *range;
+    size_t index, hi, lo;
+
+    lo = 0;
+    hi = chunk->free_ranges_count;
+
+    while (lo < hi)
+    {
+        index = lo + (hi - lo) / 2;
+        range = &chunk->free_ranges[index];
+
+        if (range->offset > offset)
+            hi = index;
+        else
+            lo = index + 1;
+    }
+
+    return lo;
+}
+
+static void vkd3d_memory_chunk_free_range(struct vkd3d_memory_chunk *chunk, const struct vkd3d_memory_allocation *allocation)
+{
+    struct vkd3d_memory_free_range *range;
```

Comment:
Do we never release chunks back to the system with vkFreeMemory? I can see a potential problem if we have fully freed chunks hanging around in a memory type which isn't usually allocated from.

Reply:
Can add that, but the reasoning (at least for DXVK where this was largely take from) is that it's very unlikely to have a fully freed chunk in the first place. I guess it might make more sense here though.

Reply:
I can see a problem with RADV since the "freed" chunks will still contribute to global BO list and potentially cause other performance issues.

Reply:
A middle ground would be to keep a maximum number of chunks alive per memory type perhaps. Could be a separate "garbage collection" after freeing a range and the chunk is now fully available.

Reply:
I went with just unconditionally destroying the chunk. Reason being that there are quite a large number of conditions to meet for a chunk to be considered for suballocation, and we probably wouldn't want to keep chunks alive that were allocated on system memory when we really wanted device memory etc.